docker/go-control-plane/fallback-control-plane.go

```diff
@@ -86,45 +80,48 @@ func (srv *controlService) StopOnRequest(_ context.Context, req *xdsconfigpb.Sto
 	}
 	res := xdsconfigpb.StopOnRequestResponse{}
 	for t, names := range srv.filters {
-		for name, _ := range names {
+		for name := range names {
 			res.Filters = append(res.Filters, &xdsconfigpb.StopOnRequestResponse_ResourceFilter{ResourceType: t, ResourceName: name})
 		}
 	}
 	return &res, nil
 }
 
-// UpsertResources allows the test to provide a new or replace existing xDS
-// resource. Notification will be sent to any control plane clients watching
+// SetResources allows the test to provide a new or replace existing xDS
+// resources. Notification will be sent to any control plane clients watching
 // the resource being updated.
-func (srv *controlService) UpsertResources(_ context.Context, req *xdsconfigpb.UpsertResourcesRequest) (*xdsconfigpb.UpsertResourcesResponse, error) {
+func (srv *controlService) SetResources(_ context.Context, req *xdsconfigpb.SetResourcesRequest) (*xdsconfigpb.SetResourcesResponse, error) {
 	srv.mu.Lock()
 	defer srv.mu.Unlock()
-	srv.version++
-	listener := controlplane.ListenerName
-	if req.Listener != nil {
-		listener = *req.Listener
+	if len(req.Resources) > 0 {
+		srv.version++
+	}
+	for _, resource := range req.Resources {
+		key := resourceKey{ResourceType: resource.Type, ResourceName: resource.Name}
+		contents := resource.Body
```

Comment:
Optional:

```golang
if contents := resource.Body; contents != nil {
	body, err := contents.UnmarshalNew()
	if err != nil {
		log.Printf("Failed to parse %s/%s: %v", key.ResourceType, key.ResourceName, err)
		continue
	}
	srv.resources[key] = &body
} else {
	delete(srv.resources, key)
}
```

Or

```golang
for _, resource := range req.Resources {
	key := resourceKey{ResourceType: resource.Type, ResourceName: resource.Name}
	contents := resource.Body
	if contents == nil {
		delete(srv.resources, key)
		continue
	}
	body, err := contents.UnmarshalNew()
	if err != nil {
		log.Printf("Failed to parse %s/%s: %v", key.ResourceType, key.ResourceName, err)
		continue
	}
	srv.resources[key] = &body
}
```

I prefer the second approach. It feels more idiomatic Go where we avoid `else` branches wherever possible and the code is less indented.

Reply:
Why do you feel that `continue` is too prone to errors? I can buy that for `goto` (or its equivalent in Go which is `break <Label>`.

Reply:
It's personal preference. I feel like
```
if (error) {
   doBad();
} else {
   doGood();
}
```
is safer then
```
if (error) {
  doBad();
  continue;
}
doGood();
```
because missing the `continue` runs `doGood()` . It is quite similar to me to fallthrough in switch/case, that is well known to cause bugs.

Reply:
This is like saying:

```
if (error) {
   doBad();
} else {
   doGood();
}
return
```

is safer than

```
if (error) {
   doBad();
   return
}
doGood();
return
```

Reply:
As I said, this is a personal preference... Thanks for the review!